pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/design/AbstractCounterCheckRule.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.rule.design;
+
+import static net.sourceforge.pmd.properties.constraints.NumericConstraints.positive;
+
+import java.lang.reflect.Modifier;
+
+import net.sourceforge.pmd.lang.apex.ast.AbstractApexNodeBase;
+import net.sourceforge.pmd.lang.apex.ast.ApexNode;
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexRule;
+import net.sourceforge.pmd.lang.rule.internal.CommonPropertyDescriptors;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+
+
+/**
+ * Abstract class for rules counting some integer metric on some node.
+ *
+ * @author Clément Fournier
+ * @since 6.7.0
+ */
+abstract class AbstractCounterCheckRule<T extends ApexNode<?>> extends AbstractApexRule {
+
+
+    private final PropertyDescriptor<Integer> reportLevel =
+        CommonPropertyDescriptors.reportLevelProperty()
+                                 .desc("Threshold above which a node is reported")
+                                 .require(positive())
+                                 .defaultValue(defaultReportLevel()).build();
+
+
+    AbstractCounterCheckRule(Class<T> nodeType) {
+        definePropertyDescriptor(reportLevel);
+        if (!(Modifier.isAbstract(nodeType.getModifiers()) || nodeType.isInterface())) {
+            addRuleChainVisit(nodeType);
+        } else {
+            assert false : "Rule chain visits must be concrete node types";
```

Comment:
why do you favor asserts? They have a number of shortcomings:

1. They are disabled by default (you would have to manually enable them passing args to Java).
2. Without them enabled, this ends in a completely unrelated `ClassCastException` at line 63…
3. … which will occur once per file, as `Exception`s are wrapped in `PMDException` during analysis and each file skipped from this rule. Such exceptions get logged to result XMLs

On the other hand, you could just throw `new AssertionError("Rule chain visits must be concrete node types")`, which will work consistently regardless of Java flags, and being an `Error`, will abort execution altogether.

Reply:
You're right, I was being lazy